Wrapping problematic fields for signatures (#2189)
httpwg/http-extensions: draft-ietf-httpbis-message-signatures.md

Note that the value for `key="c"` has been re-serialized according to the strict `member_value` algorithm.

### Binary-wrapped HTTP Fields {#http-header-byte-sequence}

> If the value of the the HTTP field in question is known by the application to cause problems with serialization, particularly with combination of multiple values as discussed in {{security-non-list}}, the signer MAY include the `bs` parameter in a component identifier to indicate the values of the fields need to be wrapped as binary structures before being combined.

Reviewer: This is not a MAY, it's at least a SHOULD because not using it would result in a security issue.

Author: I'm actually fine with this being a SHOULD for the reasons given.

```suggestion
If the value of the the HTTP field in question is known by the application to cause problems with serialization, particularly with combination of multiple values as discussed in {{security-non-list}}, the signer SHOULD include the `bs` parameter in a component identifier to indicate the values of the fields need to be wrapped as binary structures before being combined.
```

> If the value of the the HTTP field in question is known by the application to cause problems with serialization, particularly with combination of multiple values as discussed in {{security-non-list}}, the signer MAY include the `bs` parameter in a component identifier to indicate the values of the fields need to be wrapped as binary structures before being combined.

Reviewer: If we are aware of other common fields that need to be handled this way, let's list them in {{security-non-list}} as well.

Author: Set-Cookie is the only common one that I found that fits this, but I didn't want the language to preclude other weirdness out there.

If this parameter is included with a component identifier, the component value is calculated using the following algorithm:

0. Let the input be the ordered set of values for a field
1. For each field value in the set:
    0. Strip leading and trailing whitespace from each item in the list. Note that since HTTP field values are not allowed to contain leading and trailing whitespace, this will be a no-op in a compliant implementation.
    1. Remove any obsolete line-folding within the line and replace it with a single space (" "), as discussed in {{Section 5.2 of HTTP1}}. Note that this behavior is specific to {{HTTP1}} and does not apply to other versions of the HTTP specification.

>     2. Encode the string as a Byte Sequence

Reviewer: What does this mean? Maybe: encode the list as a structured field Byte Sequence using the list's ASCII representation. (This assumes that only ASCII is allowed, but that's really part of the problem.)

Author: @mnot is this an accurate way to say "take the header value as a binary"? To create the examples I encoded the strings, and field values are already a subset of ASCII characters allowed.

```suggestion
    2. Encode the field value's ASCII representation as a Byte Sequence
```

When a request message results in a signed response message, the signer can include portions of the request message in the signature base by adding the `req` parameter to the component identifier.

`req`

> : A boolean flag indicatin that the component value is derived from the request that triggered this response message and not from the response message directly.

Author: ```suggestion
: A boolean flag indicating that the component value is derived from the request that triggered this response message and not from the response message directly.
```
